File: frontend/src/services/teamChange.ts
import api from './api'

export type TeamChangeRequestStatus = 'pending' | 'approved' | 'rejected'

export interface TeamChangeRequest {
  id: string
  participant_id: string
  current_team_name?: string
  requested_team_name: string
  status: TeamChangeRequestStatus
  reason?: string
  reviewed_by?: string
  reviewed_at?: string
  created_at: string
  updated_at: string
  participant_name?: string
  league_id?: string
  reviewer_name?: string
}

export interface CreateTeamChangeRequest {
  requested_team_name: string
  reason?: string
}

export interface ReviewTeamChangeRequest {
  status: 'approved' | 'rejected'
  reason?: string
}

export interface TeamChangeRequestListResponse {
  requests: TeamChangeRequest[]
  total: number
}

export const STATUS_LABELS: Record<TeamChangeRequestStatus, string> = {
  pending: '대기 중',
  approved: '승인됨',
  rejected: '거절됨',
}

export const teamChangeService = {
  // Create a team change request (authenticated user)
  async create(leagueId: string, data: CreateTeamChangeRequest): Promise<TeamChangeRequest> {
    const response = await api.post<TeamChangeRequest>(`/leagues/${leagueId}/team-change-requests`, data)
    return response.data
  },

  // List my team change requests (authenticated user)
  async listMyRequests(leagueId: string): Promise<TeamChangeRequestListResponse> {
    const response = await api.get<TeamChangeRequestListResponse>(`/leagues/${leagueId}/my-team-change-requests`)
    return response.data
  },

  // Review (approve/reject) a team change request (director)
  async review(leagueId: string, requestId: string, data: ReviewTeamChangeRequest): Promise<void> {
    await api.put(`/leagues/${leagueId}/team-change-requests/${requestId}`, data)
  },

  // Cancel a team change request (owner, pending only)
  async cancel(leagueId: string, requestId: string): Promise<void> {
    await api.delete(`/leagues/${leagueId}/team-change-requests/${requestId}`)
  },

  // Admin: List team change requests by league
  async listByLeague(leagueId: string, status = ''): Promise<TeamChangeRequestListResponse> {
    const params = status ? `?status=${status}` : ''
    const response = await api.get<TeamChangeRequestListResponse>(`/admin/leagues/${leagueId}/team-change-requests${params}`)
    return response.data
  },
}

export default teamChangeService

File: backend/internal/model/team_change.go
package model

import (
	"time"

	"github.com/google/uuid"
)

// TeamChangeRequestStatus represents the status of a team change request
type TeamChangeRequestStatus string

const (
	TeamChangeStatusPending  TeamChangeRequestStatus = "pending"
	TeamChangeStatusApproved TeamChangeRequestStatus = "approved"
	TeamChangeStatusRejected TeamChangeRequestStatus = "rejected"
)

// TeamChangeRequest represents a team change request
type TeamChangeRequest struct {
	ID                uuid.UUID               `json:"id"`
	ParticipantID     uuid.UUID               `json:"participant_id"`
	CurrentTeamName   *string                 `json:"current_team_name,omitempty"`
	RequestedTeamName string                  `json:"requested_team_name"`
	Status            TeamChangeRequestStatus `json:"status"`
	Reason            *string                 `json:"reason,omitempty"`
	ReviewedBy        *uuid.UUID              `json:"reviewed_by,omitempty"`
	ReviewedAt        *time.Time              `json:"reviewed_at,omitempty"`
	CreatedAt         time.Time               `json:"created_at"`
	UpdatedAt         time.Time               `json:"updated_at"`

	// Joined fields
	ParticipantName *string    `json:"participant_name,omitempty"`
	LeagueID        *uuid.UUID `json:"league_id,omitempty"`
	ReviewerName    *string    `json:"reviewer_name,omitempty"`
}

// ParticipantTeamHistory represents a record of team membership
type ParticipantTeamHistory struct {
	ID              uuid.UUID  `json:"id"`
	ParticipantID   uuid.UUID  `json:"participant_id"`
	TeamName        string     `json:"team_name"`
	EffectiveFrom   time.Time  `json:"effective_from"`
	EffectiveUntil  *time.Time `json:"effective_until,omitempty"`
	ChangeRequestID *uuid.UUID `json:"change_request_id,omitempty"`
	CreatedAt       time.Time  `json:"created_at"`
}

// CreateTeamChangeRequest represents a request to create a team change request
type CreateTeamChangeRequest struct {
	RequestedTeamName string  `json:"requested_team_name" validate:"required,max=100"`
	Reason            *string `json:"reason,omitempty"`
}

// ReviewTeamChangeRequest represents a request to review (approve/reject) a team change request
type ReviewTeamChangeRequest struct {
	Status TeamChangeRequestStatus `json:"status" validate:"required"`
	Reason *string                 `json:"reason,omitempty"`
}

// TeamChangeRequestListResponse represents the response for listing team change requests
type TeamChangeRequestListResponse struct {
	Requests []*TeamChangeRequest `json:"requests"`
	Total    int                  `json:"total"`
}
